src/ExtUvLoop.php

```diff
@@ -0,0 +1,304 @@
+        $this->timers->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            (int)ceil($interval * 1000),
+            (int)ceil($interval * 1000),
+            $callback
+        );
+
+        return $timer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cancelTimer(TimerInterface $timer)
+    {
+        if (isset($this->timers[$timer])) {
+            @\uv_timer_stop($this->timers[$timer]);
+            $this->timers->detach($timer);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function futureTick($listener)
+    {
+        $this->futureTickQueue->add($listener);
+    }
+
+    public function addSignal($signal, $listener)
+    {
+        $this->signals->add($signal, $listener);
+
+        if (!isset($this->signalEvents[$signal])) {
+            $signals = $this->signals;
+            $this->signalEvents[$signal] = \uv_signal_init($this->uv);
+            \uv_signal_start($this->signalEvents[$signal], function () use ($signals, $signal) {
+                $signals->call($signal);
+            }, $signal);
+        }
+    }
+
+    public function removeSignal($signal, $listener)
+    {
+        $this->signals->remove($signal, $listener);
+
+        if (isset($this->signalEvents[$signal]) && $this->signals->count($signal) === 0) {
+            \uv_signal_stop($this->signalEvents[$signal]);
+            unset($this->signalEvents[$signal]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function run()
+    {
+        $this->running = true;
+
+        while ($this->running) {
+            $this->futureTickQueue->tick();
+
+            $hasPendingCallbacks = !$this->futureTickQueue->isEmpty();
+            $wasJustStopped = !$this->running;
+            $nothingLeftToDo = !$this->readStreams
+                && !$this->writeStreams
+                && !$this->timers->count()
+                && $this->signals->isEmpty();
+
+            // Use UV::RUN_ONCE when there are only I/O events active in the loop and block until one of those triggers,
+            // otherwise use UV::RUN_NOWAIT.
+            // @link http://docs.libuv.org/en/v1.x/loop.html#c.uv_run
+            $flags = \UV::RUN_ONCE;
+            if ($wasJustStopped || $hasPendingCallbacks) {
+                $flags = \UV::RUN_NOWAIT;
+            } elseif ($nothingLeftToDo) {
+                break;
+            }
+
+            \uv_run($this->uv, $flags);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stop()
+    {
+        $this->running = false;
+    }
+
+    private function addStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            $this->streamEvents[(int) $stream] = \uv_poll_init_socket($this->uv, $stream);
+        }
+
+        // Run in tick or else things epically fail with loop->watchers[w->fd] == w
+        $this->futureTick(function () use ($stream) {
+            $this->pollStream($stream);
```

Comment:
I'd like to have a better understanding of why this is needed and what issue this addresses. Is this an upstream issue? Could this possibly be related to https://github.com/bwoebi/php-uv/issues/49?

It's my understanding that we have a single `uv_poll_init_socket()` for a single stream with both a read and write side and may call `uv_poll_start()` multiple times on this same handle and will call `uv_poll_stop()` once both sides have been removed. This is covered by the `libuv` documentation as well as `ext-uv`, so I'm failing to see a problem here and how deferring this to the next tick could address any issues?

Reply:
Hmm I'll have a look at doing it that way. This way turned out to be the working solution after hours of debugging. Might have missed that approach 👍

Reply:
With future tick: https://travis-ci.org/reactphp/event-loop/jobs/427051253#L1898
Without future tick: https://travis-ci.org/reactphp/event-loop/jobs/427051579#L1292

Going to dive deeper into the why and see if there is an alternative solution but with a future tick it works.